### ISSUE TITLE
Teleoperate using PD_EE_Pose Controller for data collection

### ISSUE THREAD

**Issue by cublicma** [arth-shukla/mshab#50]
Hi Arth, I want to collect some data to fine-tune those pre-trained policies inside the same MSHAB environments. Firstly, do you have such scripts to do so? I'm using a modified demo_manual_control_continuous.py from Maniskill to teleoperate the Fetch robot to collect the action data. Given that MSHAB envs use PD_Joint_Delta_Pos controller, I store the 13-dim action by calculating the delta joint movement between two steps. However, during replay in the MSHAB env, I found the robot's real movement is pretty small, which mainly leads to the scale issues. I set the scale for arm and base to 0.05. The replay gets much better, but still not accurate enough for use. Can you give some hints for my methods and implementation? I would also appreciate it if you can share the data collection scripts you used to train those policies with us. Thank you.

**Comment by arth-shukla** [comment]
@cublicma Thanks for asking!

Re: finetuning policies, please refer to [the corresponding section in the README](https://github.com/arth-shukla/mshab/blob/65f09d3261e94d0d08910c2439afb1e0b0d2200e/README.md?plain=1#L215).

Re: teleop/pd_joint_delta_pos, I'm not quite sure I understand the problem. My understanding is you're looking to collect data from teleop (via e.g. ee pose controller), convert to pd_joint_delta_pos, then use this for training? If so, the steps are as follows:
1. **pd_ee_[controller_type] → pd_joint_pos**: you can get the (unnormalized) joint targets from the IK in the controller class [here](https://github.com/haosulab/ManiSkill/blob/17121e3f96e3ee3ed0c03610b17f8bc2864617af/mani_skill/agents/controllers/pd_ee_pose.py#L103).
2. **pd_joint_pos → pd_joint_delta_pos**: to get (unnormalized) delta pos control actions, simply do agent.qpos - target_qpos 

To use these as action labels, joint targets / delta targets should be clipped and scaled as [in the joint pos controller](https://github.com/haosulab/ManiSkill/blob/17121e3f96e3ee3ed0c03610b17f8bc2864617af/mani_skill/agents/controllers/pd_joint_pos.py#L74).

Re: data collection, [`scripts/gen_dataset.sh`](https://github.com/arth-shukla/mshab/blob/main/scripts/gen_dataset.sh) contains the data generation commands to generate the released dataset with the RL policies.

**Comment by cublicma** [comment]
Thanks for your quick reply! You got my question correctly and I'm indeed following the two steps you mentioned. I may take a careful look at the clipping and scaling part in the joint pos controller. Some other controllers usually have a fixed scale factor for their PD algorithm, but for Maniskill, it looks like those controllers are using a different way (clip as scale). Do I understand correctly?

Re: data collection script, since I want to define some new tasks with new objects, those pre-trained RL policies don't work for the automatic action data generation process.

**Comment by arth-shukla** [comment]
In ManiSkill for delta joint pos, the actions are clipped [-1, 1], then scaled to min/max delta pos (for fetch, [this is 0.1](https://github.com/haosulab/ManiSkill/blob/17121e3f96e3ee3ed0c03610b17f8bc2864617af/mani_skill/agents/robots/fetch/fetch.py#L134)). To get actions from target delta joint pos, just need to do the opposite.

For data collection, I don't think there are teleop data collection scripts available yet (outside of the demo_random_control scripts). For mshab we only record data from the RL policies, not from teleop.

**Comment by cublicma** [comment]
I implemented the action recording function based on your suggestion but reversed the order:

    delta = target_qpos - q_curr

However, there's still a gap during the replay ([video](https://drive.google.com/file/d/1YES_EAA2z3EH3EWuGNbcEEo2CkVdp4eh/view?usp=sharing)). The arm mismatches the trajectory during teleoperation; the gripper and base movement seem to be good. Some guess:
1. Does gravity matter? Should I tune it?
2. Calling env.step() once won't take the robot to the target_qpos due to PD. Should I take care of it?
3. Any other hints?

I can't use RL methods without much effort since the reward function is pretty hard to define for my tasks.

Pseudo code clip from my action recording function:

    q_curr = obs['agent']['qpos'][0]
    obs, rew, terminated, truncated, info = super().step(action)
    q_after = obs['agent']['qpos'][0]

    controllers = self.base_env.agent.controllers["pd_ee_delta_pose"].controllers
    target_qpos_arm = controllers["arm"]._target_qpos[0]
    target_qpos_gripper = controllers["gripper"]._target_qpos[0]
    target_qpos_body = controllers["body"]._target_qpos[0]

    target_qpos_full = q_curr.copy()
    target_qpos_full[1] = target_qpos_body[0]  # head_pan_joint
    target_qpos_full[3] = target_qpos_body[1]  # head_tilt_joint
    target_qpos_full[0] = target_qpos_body[2]  # torso_lift_joint
    target_qpos_full[2] = target_qpos_arm[0]  # shoulder_pan_joint
    target_qpos_full[4] = target_qpos_arm[1]  # shoulder_lift_joint
    target_qpos_full[5] = target_qpos_arm[2]  # upperarm_roll_joint
    target_qpos_full[6] = target_qpos_arm[3]  # elbow_flex_joint
    target_qpos_full[7] = target_qpos_arm[4]  # forearm_roll_joint
    target_qpos_full[8] = target_qpos_arm[5]  # wrist_flex_joint
    target_qpos_full[9] = target_qpos_arm[6]  # wrist_roll_joint
    target_qpos_full[10] = target_qpos_gripper[1]  # r_gripper_finger_joint
    target_qpos_full[11] = target_qpos_gripper[0]  # l_gripper_finger_joint

    # Compute deltas for joint-delta actions (gripper is handled as absolute later)
    delta = target_qpos_full - q_curr

    # Normalize and clip
    arm_in = np.clip(delta[IDX["arm"]] / self._arm_scale[0], -1.0, 1.0)
    head_in = np.clip(delta[IDX["head"]] / self._head_scale, -1.0, 1.0)
    torso_in = np.clip(delta[IDX["torso"]][0] / self._torso_scale, -1.0, 1.0)

    # Gripper: store absolute, use post-step state for visibility
    grip_pos = target_qpos_full[10]

    # Base velocities from base_link.pose
    x0, y0, yaw0 = self._prev_base
    x1, y1, yaw1 = self.read_base_pose(self.base_env)
    dx, dy = x1 - x0, y1 - y0
    v_fwd = (np.cos(yaw0) * dx + np.sin(yaw0) * dy) / self._control_dt   # 1.0 / float(self.unwrapped.control_freq)
    dyaw = (yaw1 - yaw0 + np.pi) % (2 * np.pi) - np.pi
    w_z = dyaw / self._control_dt
    base_2 = [v_fwd, w_z]

    action = np.concatenate([arm_in, [grip_pos], head_in, [torso_in], base_2], axis=0)

**Comment by arth-shukla** [comment]
@cublicma Apologies for missing your message! For some reason, I didn't get a GitHu notification...

In any case, I've attached a script below which translates pd_ee_delta_pose to pd_joint_pos and pd_joint_delta_pos: [demo_manual_control.py](https://github.com/user-attachments/files/22170433/demo_manual_control.py)

A few notes:
- Only the arm actions need to be translated. The base, body, and gripper controllers are always the same.
- The below code is using the current [latest commit from ManiSkill](https://github.com/haosulab/ManiSkill/commit/47998f619190439e0be3f6b3a5078de2593cfc1f). There are some [IK fixes](https://github.com/haosulab/ManiSkill/pull/996) on main which I'll merge into the mshab branch prob after the next ManiSkill beta release.
- The below vids use CPU sim. If you're going to train policies on GPU sim, then you should use GPU sim for data collection due to differences in the CPU/GPU physics backends.

Eg commands and corresponding video replay outputs are as follows:

To generate example data:
```
python mani_skill/examples/demo_manual_control.py -e PickCube-v1 robot_uids fetch
```

---

To replay pd_ee_delta_pose actions:
```
python mani_skill/examples/demo_manual_control.py -e PickCube-v1 -c pd_ee_delta_pose --record-dir out --load-actions robot_uids fetch
```
Corresponding video:
<video src="https://github.com/user-attachments/assets/5a98edb2-6560-4bcd-b6ca-20bfa8ed0643" />

---

To replay pd_joint_pos actions:
```
python mani_skill/examples/demo_manual_control.py -e PickCube-v1 -c pd_joint_pos --record-dir out --load-actions robot_uids fetch
```
Corresponding video:
<video src="https://github.com/user-attachments/assets/b203da4e-6342-45e8-b4fe-aa5dcd638e5d" />

---

To replay pd_joint_delta_pos actions:
```
python mani_skill/examples/demo_manual_control.py -e PickCube-v1 -c pd_joint_delta_pos --record-dir out --load-actions robot_uids fetch
```
Corresponding video:
<video src="https://github.com/user-attachments/assets/f48a5ef1-a6b5-447e-bb84-6ec5b2e0d014" />
